### ISSUE TITLE
caffe2fluid直接保存inference_model 的模型输出和使用load_var将模型载入，并使用save_inference_model保存输出存在差异

### ISSUE THREAD

**Issue by miraclebiu** [PaddlePaddle/Paddle#17697]
1）PaddlePaddle版本：1.4.1
    
    3）GPU：k40、CUDA8和CUDNN7
    4）系统环境：centos python2.7
- 训练信息
    1）单机，多卡
    2）k40
    3）Operator信息
- 复现信息：如为报错，请给出复现环境、复现步骤
- 问题描述：请详细描述您的问题，同步贴出报错信息、日志、可复现的代码片段

使用caffe2fluid官方提供的转换代码，将npy文件利用save_inference_model保存权重结构，模型在本地测试可以正常输出结果，之后为了finetune方便，将save_inference_model接口改为save_persistable保存，并将网络结构中的is_test均设置为True，使用load_var进行加载权重（为了避免bn层的moment找不到），设置sgd，**学习率为0**（这里的学习率为0是为了将后续流程跑通），最后使用save_inference_model保存权重结构，发现两次save_inference_model 的inference（is_test=False）结果不一致

caffe2fluid中的保存模型代码为：
        
`       
        fluid.io.save_persistables(exe,dirname=fluid_path)
        #fluid.io.save_inference_model(
        #    fluid_path, [input_name],
        #    output_vars,
        #    exe,
        #    main_program=None,
        #    model_filename=model_filename,
        #    params_filename=params_filename)
`

bn层转换过来的代码为：

`
@layer
    def batch_normalization(self,
                            input,
                            name,
                            scale_offset=True,
                            eps=1e-5,
                            relu=False,
                            relu_negative_slope=0.0):
        # NOTE: Currently, only inference is supported
        fluid = import_fluid()
        prefix = name + '_' 
        param_attr = None if scale_offset is False else fluid.ParamAttr(
            name=prefix + 'scale')
        bias_attr = None if scale_offset is False else fluid.ParamAttr(
            name=prefix + 'offset')
        mean_name = prefix + 'mean'
        variance_name = prefix + 'variance'

        leaky_relu = False
        act = 'relu'
        if relu is False:
            act = None
        elif relu_negative_slope != 0.0:
            leaky_relu = True
            act = None

        output = fluid.layers.batch_norm(
            name=self.get_unique_output_name(name, 'batch_norm'),
            input=input,
            is_test=False,                                                                                            
            param_attr=param_attr,
            bias_attr=bias_attr,
            moving_mean_name=mean_name,
            moving_variance_name=variance_name,
            epsilon=eps,
            act=act)

        if leaky_relu:
            output = fluid.layers.leaky_relu(output, alpha=relu_negative_slope)

        return output`

 

dropout转换得到的代码为：
` @layer
    def dropout(self, input, drop_prob, name, is_test=False):
        fluid = import_fluid()
        if is_test:
            output = input
        else:
            output = fluid.layers.dropout(
                input,
                dropout_prob=drop_prob,
                is_test=is_test,
                name=self.get_unique_output_name(name, 'dropout'))
        return output

`

网络inference使用的代码为：
`print(each_img)
        img_test = Image.open(each_img)
        img_resize = img_test.resize((224, 224), Image.BILINEAR)
        out_img= np.array(img_resize)
        mean = [104.,117.,123.]# bgr order
        #mean = [123, 117, 104]
        std = [1,1,1]
        #img_resize = img_test.resize((224, 224), Image.BILINEAR)
        mean = np.array(mean).reshape((3, 1, 1)) 
        out_img = out_img.transpose((2,0,1))# hwc->chw
        out_img = out_img[(2,1,0),:,:]# caffe model is bgr so rgb->bgr
        out_img = out_img  - mean 
        out_img = out_img.reshape(1,3,224,224).astype('float32')
[inference_program, feed_target_names, fetch_targets] = fluid.io.load_inference_model(dirname=fluid_weights_dir, executor=exe, model_filename='model', params_filename='params')
        results = exe.run(inference_program, 
                          feed = {feed_target_names[0]: out_img},
                          fetch_list=fetch_targets)

        #print(results)
        print(results[0].argmax(),results[1].argmax(),results[2].argmax(),results[3].argmax())
        print(results[0].max(),results[1].max(),results[2].max(),results[3].max())`

        
网络载入权重代码：
`exe.run(startup_program)
    print('start loading models')
    def if_exist(var):
        var_file_name = os.path.join(net_weight_dir,var.name)
        if not os.path.exists(var_file_name):
            #print(var.name,"is not exists in weight file")
            return False
        return True
    fluid.io.load_vars(executor=exe,dirname=net_weight_dir,main_program=train_program,predicate=if_exist)
    print('load Done!') `

    

网络训练代码为：
`nb_epochs = 2 
    for ii in range(0, nb_epochs): 
    
        #for train
        for i, data in enumerate(train_reader()):
            results = exe.run(program=train_parallel_prog, fetch_list=[loss_list[0].name], feed=feeder.feed(data))
            #pdb.set_trace()
            results = np.mean(results)
            print('{} iterations, loss: {:.3f}'.format(i,results))
        model_path = "./output/fg_car_rec_v7_"+str(ii)
        #fluid.io.save_persistables(executor=exe, dirname=model_path,main_program=train_parallel_prog)
     
        fluid.io.save_inference_model(model_path,[feed_list[0].name],pred_list,exe, main_program=train_program, model_filename='model', params_filename='params')`

**Comment by heavengate** [comment]
load_var之后是经过两个epoch的训练之后再save_inference_model的，即使leaning rate=0， batch norm的moving_mean和moving_variance也会更新，会导致diff

**Comment by miraclebiu** [comment]
在bn层中设置use_global_stats=True即可
